Source/DafnyCore/Resolver/PreType/PreTypeResolve.cs

```diff
@@ -574,16 +576,22 @@ private void AddComparableTypesDefault(PreType a, PreType b) {
     /// Additionally, under the legacy option /generalNewtypes:0 (which will be phased out over time), the latter also allows
     /// several additional cases, see IsConversionCompatible.
     /// </summary>
-    bool CheckComparableTypes(PreType a, PreType b, bool allowConversion) {
+    bool CheckComparableTypes(PreType a, PreType b, bool allowConversion, bool allowCommonSuperType) {
       if (PreType.Same(a, b)) {
         // this allows the case where "a" and "b" are proxies that are equal
         return true;
       }
       if (a.Normalize() is not DPreType aa || b.Normalize() is not DPreType bb) {
         return false;
       }
-      if (IsSuperPreTypeOf(aa, bb) || IsSuperPreTypeOf(bb, aa)) {
-        return true;
+      if (allowCommonSuperType) {
+        if (HaveCommonSuperPreType(aa, bb)) {
+          return true;
+        }
+      } else {
+        if (IsSuperPreTypeOf(aa, bb) || IsSuperPreTypeOf(bb, aa)) {
+          return true;
+        }
```

Comment:
Another way to accomplish this is to leave the original comparison

``` c#
if (IsSuperPreTypeOf(aa, bb) || IsSuperPreTypeOf(bb, aa)) {
        return true;
}
```

and then do

``` c#
if (allowCommonSuperType && HaveCommonSuperPreType(aa, bb)) {
  return true;
}
```

where the initial comparison `if (IsSuperPreTypeOf(t, u) || IsSuperPreTypeOf(u, t)) ...` inside `HaveCommonSuperPreType` can be removed.